pkg/syncer/controllermanager/controllermanager.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllermanager
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	ControllerNamePrefix = "syncer-controller-manager-"
+)
+
+type InformerSource struct {
+	Subscribe func(id string) <-chan struct{}
+	Informer  func(gvr schema.GroupVersionResource) (informer cache.SharedIndexInformer, known, synced bool)
+}
+
+type Controller interface {
+	Start(ctx context.Context, numThreads int)
+}
+
+type ControllerDefintion struct {
+	RequiredGVRs []schema.GroupVersionResource
+	NumThreads   int
+	Create       func(syncedInformers map[schema.GroupVersionResource]cache.SharedIndexInformer) (Controller, error)
+}
+
+func NewControllerManager(ctx context.Context, suffix string, informerSource InformerSource, controllers map[string]ControllerDefintion) *ControllerManager {
```

Comment:
maybe I miss something: what is the added value of this controller manager concept? The controller definition are obviously static. If it is just about waiting for informer sync, create a simple helper/wrapper around a controller start func.

Reply:
disregard. https://github.com/kcp-dev/kcp/pull/2452/files#r1063304421

Reply:
I modified the ControllerManager to only start/stop the GVR-specific controllers, but not create them anymore. This makes the code simpler.
Thanks for pointing this out.

Reply:
Well, that"s not true finally. I still need to create the controller through the ControllerManager, since the controller queue is typically shot down at the end of the Start method. And I don't want to force a diverging implementation of those controllers.
However I implemented the Create support a bit more cleanly I think.